// @squircle-js/react/dist/index.js
"use strict";
"use client";
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
function _object_without_properties(source, excluded) {
    if (source == null) return {};
    var target = _object_without_properties_loose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for(i = 0; i < sourceSymbolKeys.length; i++){
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) continue;
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
            target[key] = source[key];
        }
    }
    return target;
}
function _object_without_properties_loose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for(i = 0; i < sourceKeys.length; i++){
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        target[key] = source[key];
    }
    return target;
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = function(target, all) {
    for(var name in all)__defProp(target, name, {
        get: all[name],
        enumerable: true
    });
};
var __copyProps = function(to, from, except, desc) {
    if (from && (typeof from === "undefined" ? "undefined" : _type_of(from)) === "object" || typeof from === "function") {
        var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
        try {
            var _loop = function() {
                var key = _step.value;
                if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
                    get: function() {
                        return from[key];
                    },
                    enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
                });
            };
            for(var _iterator = __getOwnPropNames(from)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true)_loop();
        } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
        } finally{
            try {
                if (!_iteratorNormalCompletion && _iterator.return != null) {
                    _iterator.return();
                }
            } finally{
                if (_didIteratorError) {
                    throw _iteratorError;
                }
            }
        }
    }
    return to;
};
var __toCommonJS = function(mod) {
    return __copyProps(__defProp({}, "__esModule", {
        value: true
    }), mod);
};
// src/index.tsx
var index_exports = {};
__export(index_exports, {
    Squircle: function() {
        return Squircle;
    },
    SquircleNoScript: function() {
        return SquircleNoScript;
    },
    StaticSquircle: function() {
        return StaticSquircle;
    }
});
module.exports = __toCommonJS(index_exports);
var import_react5 = require("react");
var import_react_slot2 = require("@radix-ui/react-slot");
var import_figma_squircle2 = require("figma-squircle");
// src/use-element-size.ts
var import_react3 = require("react");
// src/use-event-listener.ts
var import_react2 = require("react");
// src/use-isomorphic-layout-effect.ts
var import_react = require("react");
var useIsomorphicLayoutEffect = typeof window !== "undefined" ? import_react.useLayoutEffect : import_react.useEffect;
// src/use-event-listener.ts
function useEventListener(eventName, handler, element, options) {
    var savedHandler = (0, import_react2.useRef)(handler);
    useIsomorphicLayoutEffect(function() {
        savedHandler.current = handler;
    }, [
        handler
    ]);
    (0, import_react2.useEffect)(function() {
        var _element_current;
        var targetElement = (_element_current = element === null || element === void 0 ? void 0 : element.current) !== null && _element_current !== void 0 ? _element_current : window;
        if (!(targetElement === null || targetElement === void 0 ? void 0 : targetElement.addEventListener)) return;
        var listener = function(event) {
            return savedHandler.current(event);
        };
        targetElement.addEventListener(eventName, listener, options);
        return function() {
            targetElement.removeEventListener(eventName, listener, options);
        };
    }, [
        eventName,
        element,
        options
    ]);
}
// src/use-element-size.ts
function useElementSize(defaultSize) {
    var _ref = _sliced_to_array((0, import_react3.useState)(null), 2), ref = _ref[0], setRef = _ref[1];
    var _defaultSize_defaultWidth, _defaultSize_defaultHeight;
    var _ref1 = _sliced_to_array((0, import_react3.useState)({
        width: (_defaultSize_defaultWidth = defaultSize.defaultWidth) !== null && _defaultSize_defaultWidth !== void 0 ? _defaultSize_defaultWidth : 0,
        height: (_defaultSize_defaultHeight = defaultSize.defaultHeight) !== null && _defaultSize_defaultHeight !== void 0 ? _defaultSize_defaultHeight : 0
    }), 2), size = _ref1[0], setSize = _ref1[1];
    var handleSize = (0, import_react3.useCallback)(function() {
        var _ref_offsetWidth, _ref_offsetHeight;
        setSize({
            width: (_ref_offsetWidth = ref === null || ref === void 0 ? void 0 : ref.offsetWidth) !== null && _ref_offsetWidth !== void 0 ? _ref_offsetWidth : 0,
            height: (_ref_offsetHeight = ref === null || ref === void 0 ? void 0 : ref.offsetHeight) !== null && _ref_offsetHeight !== void 0 ? _ref_offsetHeight : 0
        });
    }, [
        ref === null || ref === void 0 ? void 0 : ref.offsetHeight,
        ref === null || ref === void 0 ? void 0 : ref.offsetWidth
    ]);
    useEventListener("resize", handleSize);
    useIsomorphicLayoutEffect(function() {
        handleSize();
    }, [
        ref === null || ref === void 0 ? void 0 : ref.offsetHeight,
        ref === null || ref === void 0 ? void 0 : ref.offsetWidth
    ]);
    return [
        setRef,
        size
    ];
}
// src/no-js.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var SquircleNoScript = function() {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("noscript", {
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("style", {
            type: "text/css",
            children: "[data-squircle] { clip-path: none !important; border-radius: attr(data-squircle) !important; }"
        })
    });
};
// src/StaticSquircle.tsx
var import_react4 = require("react");
var import_react_slot = require("@radix-ui/react-slot");
var import_figma_squircle = require("figma-squircle");
var import_jsx_runtime2 = require("react/jsx-runtime");
var StaticSquircle = function(_param) {
    var asChild = _param.asChild, width = _param.width, height = _param.height, cornerRadius = _param.cornerRadius, cornerSmoothing = _param.cornerSmoothing, style = _param.style, props = _object_without_properties(_param, [
        "asChild",
        "width",
        "height",
        "cornerRadius",
        "cornerSmoothing",
        "style"
    ]);
    var Component = asChild ? import_react_slot.Slot : "div";
    var path = (0, import_react4.useMemo)(function() {
        return (0, import_figma_squircle.getSvgPath)({
            width: width,
            height: height,
            cornerRadius: cornerRadius,
            cornerSmoothing: cornerSmoothing
        });
    }, [
        width,
        height,
        cornerRadius,
        cornerSmoothing
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(Component, _object_spread({
        style: _object_spread({
            clipPath: "path('".concat(path, "')")
        }, style)
    }, props));
};
// src/index.tsx
var import_jsx_runtime3 = require("react/jsx-runtime");
function Squircle(_param) {
    var cornerRadius = _param.cornerRadius, _param_cornerSmoothing = _param.cornerSmoothing, cornerSmoothing = _param_cornerSmoothing === void 0 ? 0.6 : _param_cornerSmoothing, asChild = _param.asChild, style = _param.style, w = _param.width, h = _param.height, defaultWidth = _param.defaultWidth, defaultHeight = _param.defaultHeight, props = _object_without_properties(_param, [
        "cornerRadius",
        "cornerSmoothing",
        "asChild",
        "style",
        "width",
        "height",
        "defaultWidth",
        "defaultHeight"
    ]);
    var Component = asChild ? import_react_slot2.Slot : "div";
    var _useElementSize = _sliced_to_array(useElementSize({
        defaultWidth: defaultWidth,
        defaultHeight: defaultHeight
    }), 2), ref = _useElementSize[0], _useElementSize_ = _useElementSize[1], width = _useElementSize_.width, height = _useElementSize_.height;
    var actualWidth = w !== null && w !== void 0 ? w : width;
    var actualHeight = h !== null && h !== void 0 ? h : height;
    var path = (0, import_react5.useMemo)(function() {
        if (actualWidth === 0 || actualHeight === 0) return "";
        return (0, import_figma_squircle2.getSvgPath)({
            width: actualWidth,
            height: actualHeight,
            cornerRadius: cornerRadius,
            cornerSmoothing: cornerSmoothing
        });
    }, [
        actualWidth,
        actualHeight,
        cornerRadius,
        cornerSmoothing
    ]);
    return /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(Component, _object_spread_props(_object_spread({}, props), {
        ref: ref,
        style: _object_spread_props(_object_spread({}, style), {
            borderRadius: cornerRadius,
            width: w !== null && w !== void 0 ? w : defaultWidth,
            height: h !== null && h !== void 0 ? h : defaultHeight,
            clipPath: "path('".concat(path, "')")
        }),
        "data-squircle": cornerRadius
    }));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
    Squircle: Squircle,
    SquircleNoScript: SquircleNoScript,
    StaticSquircle: StaticSquircle
});
